### ISSUE TITLE
[no-unnecessary-condition] Shows optional chaining on mapped type as unnecessary

### ISSUE THREAD

**Issue by SimonSimCity** [typescript-eslint/typescript-eslint#4954]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

To be added ...

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": ["error"]
  }
}
```

```TS
interface Property {
  [key: string]: [string] | undefined;
}

// Works correctly
((p: Property) => {
  return p.test?.[0].charAt(2);
})({} as Property);

// Should also not report anything ...
((p: Property | undefined) => {
  return p?.test?.[0].charAt(2);
})(undefined);
```

My tsconfig:

```
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es2020",
    "lib": ["es2020"],
    "outDir": "dist",
    "rootDir": ".",
    "sourceMap": true,
    "strict": true,
    "esModuleInterop": true,
    "resolveJsonModule": true
  }
}
```

**Expected Result**

Not to report any error.

**Actual Result**

The rule reports:

`Unnecessary optional chain on a non-nullish value. eslint[@typescript-eslint/no-unnecessary-condition](https://typescript-eslint.io/rules/no-unnecessary-condition)`

**Additional Info**

I've started off with older versions of the packages listed below, updated to the mentioned versions, but the problem still persists. Only `node` I haven't updated, of all packages the mentioned version is the latest as of the time writing.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.23.0` |
| `@typescript-eslint/parser`        | `5.23.0` |
| `TypeScript`                       | `4.6.4` |
| `ESLint`                           | `8.15.0` |
| `node`                             | `14.17.3` |

**Comment by bradzacher** [comment]
Please make sure you use an issue template via the chooser when filing an issue to ensure you provide all of the relevant information.

**Comment by bradzacher** [comment]
repro: https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLAID2ZArgghPtFAIaFZn60IATOEjj9UALjABtVBEKFahVAF0wIYACgNiJHIBmHNmAAKCvIWxgA3hrDSA1hCySoSYggDmyyVNfuvYAA+YEwCEHqIEAIA3BoAvlrAwGAA6or2UGB88qxIMFgaABSFOJKmtObYAJRgALwAfNa2YIQQSAyECGA4AHS6rgD8PVIADMo9+AAWXACCSIUATFWxcVWFVnFgHJnllVjLickAypO0DDACWzBQtGB0KK04iigcCFhIk4geYD2-RSVlMxySzBULhSICGoNJp2VrtTrdIb9JBDUbjKazeZLFZrMERFiQ2JAA&rules=N4XyA&tsConfig=N4IgzgLgTglgxhEAuaBXApgXyA

**Comment by evanwang0** [comment]
Here's another, simpler case. 

https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLAID2ZArgghPtFAIaFZn60IATOEjj9UALjABtVBEKFahVAF0wIYACgN2PGAAitAOZgAvGADeGsGABGhCBAGSoSYgkMBuDQF8tfBC7SAkbKkgaGUqpmkV4awYYA-AB0dg4CHkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A
